lexer.go

```diff
@@ -34,11 +38,72 @@ type Pos struct {
 	Offset int
 }
 
+type Encoding struct {
+	decoder    *encoding.Decoder
+	name       string
+	charLength func([]byte) int
+}
+
+func (e *Encoding) Enabled() bool {
+	return e.decoder != nil && e.charLength != nil
+}
+
+func NewEncoding(label string) Encoding {
+	if len(label) == 0 {
+		return Encoding{}
+	}
+	e, name := charset.Lookup(label)
+	if e == nil {
+		return Encoding{name: name}
+	}
+	return Encoding{
+		decoder:    e.NewDecoder(),
+		name:       name,
+		charLength: charset.FindNextCharacterLength(name),
+	}
+}
+
+func (e *Encoding) transformString(lit string) (string, bool) {
+	src := Slice(lit)
+	dest := make([]byte, len(src)*2)
+	var destOffset, srcOffset int
+	ok := true
+	for {
+		nextLen := 4
+		if e.charLength != nil {
+			nextLen = e.charLength(src[srcOffset:])
+		}
+		srcEnd := srcOffset + nextLen
+		if srcEnd > len(src) {
+			srcEnd = len(src)
+		}
+		nDest, nSrc, err := e.decoder.Transform(dest[destOffset:], src[srcOffset:srcEnd], false)
+		destOffset += nDest
+		srcOffset += nSrc
+		if err == nil {
+			if srcOffset >= len(src) {
+				return string(dest[:destOffset]), ok
+			}
+		} else if err == transform.ErrShortDst {
```

Comment:
Why is there no error here?

Reply:
Because this error indicates the destination bytes array is too short, we need to enlarge the capacity.